packages/ckeditor5-find-and-replace/src/findandreplaceediting.ts

```diff
@@ -269,4 +197,106 @@ export default class FindAndReplaceEditing extends Plugin {
+		const removedMarkers = new Set<string>();
+		const model = this.editor.model;
+		const { results } = this.state!;
+
+		const changes = model.document.differ.getChanges() as Array<Exclude<DiffItem, DiffItemAttribute>>;
+		const changedMarkers = model.document.differ.getChangedMarkers();
+
+		// Get nodes in which changes happened to re-run a search callback on them.
+		changes.forEach( change => {
+			if ( !change.position ) {
+				return;
+			}
+
+			if ( change.name === '$text' || ( change.position.nodeAfter && model.schema.isInline( change.position.nodeAfter ) ) ) {
+				changedNodes.add( change.position.parent as Element );
+
+				[ ...model.markers.getMarkersAtPosition( change.position ) ].forEach( markerAtChange => {
+					removedMarkers.add( markerAtChange.name );
+				} );
+			} else if ( change.type === 'insert' && change.position.nodeAfter ) {
+				changedNodes.add( change.position.nodeAfter );
+			}
+		} );
+
+		// Get markers from removed nodes also.
+		model.document.differ.getChangedMarkers().forEach( ( { name, data: { newRange } } ) => {
+			if ( newRange && newRange.start.root.rootName === '$graveyard' ) {
+				removedMarkers.add( name );
+			}
+		} );
+
+		// Get markers from the updated nodes and remove all (search will be re-run on these nodes).
+		changedNodes.forEach( node => {
+			const markersInNode = [ ...model.markers.getMarkersIntersectingRange( model.createRangeIn( node as Element ) ) ];
+
+			markersInNode.forEach( marker => removedMarkers.add( marker.name ) );
+		} );
+
+		// Remove results from the changed part of content.
+		removedMarkers.forEach( markerName => {
+			if ( !results.has( markerName ) ) {
+				return;
+			}
+
+			if ( results.get( markerName ) === this.state!.highlightedResult ) {
+				this.state!.highlightedResult = null;
+			}
+
+			results.remove( markerName );
+		} );
+
+		// Run search callback again on updated nodes.
+		const changedSearchResults: Array<ResultType> = [];
+		const findAndReplaceUtils: FindAndReplaceUtils = this.editor.plugins.get( 'FindAndReplaceUtils' );
+
+		changedNodes.forEach( nodeToCheck => {
+			const changedNodeSearchResults = findAndReplaceUtils.updateFindResultFromRange(
+				model.createRangeOn( nodeToCheck ), model, this.state!.lastSearchCallback!, results
+			);
+
+			changedSearchResults.push( ...changedNodeSearchResults );
+		} );
+
+		changedMarkers.forEach( markerToCheck => {
+			// Handle search result highlight update when T&C plugin is active.
+			// Lookup is performed only on newly inserted markers.
+			if ( markerToCheck.data.newRange ) {
+				const changedNodeSearchResults = findAndReplaceUtils.updateFindResultFromRange(
+					markerToCheck.data.newRange, model, this.state!.lastSearchCallback!, results
+				);
+
+				console.info( 'new', [ ...markerToCheck.data.newRange ] );
+
+				changedSearchResults.push( ...changedNodeSearchResults );
+
+				console.info( 'new found', changedNodeSearchResults );
+			} else if ( markerToCheck.data.oldRange &&
```

Comment:
I think this functionality will be difficult to implement and it might not be possible to do it here (in `change` event callback). We do know about `oldRange` but it was the range before any change happened. The model might have changed greatly, and the `oldRange` might be quite outdated and point to incorrect paths.

Which you already observed, but as we've seen, it crashed in another scenario and it still doesn't work other scenarios.

I'd drop this requirement for now. That is, do not react to deletion suggestion marker changes (or do not react to deletion suggestion markers being removed, as it seems it is easy to react to such markers being added). Let's move this to a separate ticket.

Other than that it seems to me that this whole callback would use some refactor but let's close what we have since the QA tests went fine.

Reply:
We can add followup for refactor as well, although I don't assume that we will find time for it.

Reply:
I added your comment to newly created issue about this problem.